httpbp/server.go

```diff
@@ -0,0 +1,271 @@
+package httpbp
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"net/http/httptest"
+	"sync"
+
+	baseplate "github.com/reddit/baseplate.go"
+	"github.com/reddit/baseplate.go/batcherror"
+)
+
+// EndpointRegistry is the minimal interface needed by a Baseplate HTTP server for
+// the underlying HTTP server.
+//
+// *http.ServeMux implements this interface and is the default EndpointRegistry
+// used by NewBaseplateServer.
+type EndpointRegistry interface {
+	http.Handler
+
+	Handle(pattern string, handler http.Handler)
+}
+
+var (
+	_ EndpointRegistry = (*http.ServeMux)(nil)
+)
+
+type httpHandlerFactory struct {
+	middlewares []Middleware
+}
+
+func (f httpHandlerFactory) NewHandler(name string, handle HandlerFunc, middlewares ...Middleware) http.Handler {
+	wrappers := make([]Middleware, 0, len(f.middlewares)+len(middlewares))
+	wrappers = append(wrappers, f.middlewares...)
+	wrappers = append(wrappers, middlewares...)
+	return NewHandler(name, handle, wrappers...)
+}
+
+// Pattern is the pattern passed to a EndpointRegistry when registering an
+// Endpoint.
+type Pattern string
+
+// Endpoint holds the values needed to create a new HandlerFunc.
+type Endpoint struct {
+	// Name is required, it is the "name" of the endpoint that will be passed
+	// to any Middleware wrapping the HandlerFunc.
+	Name string
+
+	// Handle is required, it is the base HandlerFunc that will be wrapped
+	// by any Middleware.
+	Handle HandlerFunc
+
+	// Middlewares is an optional list of additional Middleware to wrap the
+	// given HandlerFunc.
+	Middlewares []Middleware
+}
+
+// Validate checks for input errors on the Endpoint and returns an error
+// if any exist.
+func (e Endpoint) Validate() error {
+	var err batcherror.BatchError
+	if e.Name == "" {
+		err.Add(errors.New("httpbp: Endpoint.Name must be non-empty"))
+	}
+	if e.Handle == nil {
+		err.Add(errors.New("httpbp: Endpoint.Handle must be non-nil"))
+	}
+	return err.Compile()
+}
+
+// ServerArgs defines all of the arguments used to create a new HTTP
+// Baseplate server.
+type ServerArgs struct {
+	// Baseplate is a required argument to NewBaseplateServer and must
+	// be non-nil.
+	Baseplate baseplate.Baseplate
+
+	// Endpoints is the mapping of endpoint patterns to Endpoint objects that
+	// the Server will handle.
+	//
+	// While endpoints is not technically required, if none are provided, your
+	// server will not handle any Endpoints.
+	Endpoints map[Pattern]Endpoint
```

Comment:
🔕 Looking at [the doc of `http.ServeMux`](https://pkg.go.dev/net/http?tab=doc#ServeMux), it handles pattern conflicting in a deterministic way, not depending on the order of them being registered, so using a map is fine (as the iteration of a map is non-deterministic in go). But if the user uses a different mux implementation that depending on the order of endpoints are registered, using a map will cause non-deterministic result. Maybe add `Pattern` as a field of `Endpoint` and make this a slice of `Endpoint`s is better. But I don't feel strongly (using map can make sure that we don't have 2 exactly same patterns, which is a case that would panic `http.ServeMux`, so that's a pro for using a map.)

Reply:
I think that's a _reasonable_ approach too (we could always update `ValidateAndSetDefaults` to check for duplicate patterns as well).

Reply:
@konradreiche do you have any feelings on the matter?  The _entire_ reason I made an interface to plug this in is so we could potentially support things other than `http.ServeMux` so it might make sense.

Reply:
I feel that the way `http.ServeMux` handles conflicts, as described in its doc, makes a lot of sense so I would imagine most of the actual good mux implementations should follow it to some level, as a result map is likely to work fine (if it depends on the order, we might not want to use that mux implementation in the first place?)

Reply:
WFM